srml/contract/src/account_db.rs

```diff
@@ -62,7 +63,11 @@ pub trait AccountDb<T: Trait> {
 pub struct DirectAccountDb;
 impl<T: Trait> AccountDb<T> for DirectAccountDb {
 	fn get_storage(&self, _account: &T::AccountId, trie_id: Option<&TrieId>, location: &StorageKey) -> Option<Vec<u8>> {
-		trie_id.and_then(|id| child::get_raw(id, location))
+		// TODO pass optional SubTrie or change def to use subtrie (put the subtrie in cache (rc one of
```

Comment:
I did use account_db trie_id (see other comments) as keyspace (and as subtrie parent location).
So there is a need for caching SubTrie struct. TODO issue for that ?
There is a possible optimization by putting SubTrie directly at Account_id location, this would only be possible by:
- storing account infos at the same location as subtrie (in subtrie prefix)
- changing this pr to allow storing subtrie at any location
For both point it requires implementing a way to store subtrie with different encoder (and with additional info).

Reply:
> So there is a need for caching SubTrie struct. TODO issue for that ?

hmm I can also think of another anwser. The `trie_id: Option<&TrieId>` is already an optimisation that says: if there is a trie_id in direct_storage then you have to give me that I won't do the lookup from AccountId, so you can cache it for me.
This thoses changes you introduce we can just say: if there is a subtrie associated to the account already then give it I won't do the lookup.

Having subtrie in AccoundInfo seems cool though but not mandatory here.

> @thiolliere maybe this pr (not sure it will get merged there might be better way of fixing the collision), could be of interest regarding 1882 or 1883 .

I don't in which context this collision can happen I design things without taking this into consideration though. But interesting thanks

Reply:
Yes I see your optim, it is better design (that is the reason why I pinged you :)

> I don't in which context this collision can happen I design things without taking this into consideration though. But interesting thanks

without this pr it happens whenever two subtries are similar and content got pruned (for contract it should happen a lot). It only requires that both subtries got same branch path (also content (same proof would be more correct)) to the deleted value.

> Having subtrie in AccoundInfo seems cool though but not mandatory here.

As long as we use a trie_id as subtrie key from account_id (merging account info and trie info) there is no need to take this into consideration.